replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/commit/AsyncPeriodicTopicCommit.scala

```diff
@@ -0,0 +1,204 @@
+    RebalanceCallback.lift(stateRef.update(_.removeAssignedPartitions(partitions.toSortedSet)))
+  }
+
+  // opened to the package for tests
+  private[commit] def periodicCommitStep(): F[Unit] = {
+    for {
+      state          <- stateRef.get
+      offsetsToCommit = state.partitionOffsetsToCommit
+
+      anythingCommitted <- commitIfNotEmpty[F](offsetsToCommit, commitF = consumer.commitLater)
+        .recoverWith {
+          case _: RebalanceInProgressException =>
+            Applicative[F].pure(false) // ignore, commit will be retried on the next step
+        }
+
+      _ <-
+        if (anythingCommitted) {
+          stateRef.update(_.applyCommittedForAssigned(offsetsToCommit)) >>
+            log.debug(s"periodic offset commit success: $offsetsToCommit")
+        } else Applicative[F].unit
+    } yield ()
+  }
+
+  private def commitIfNotEmpty[G[_]: MonadThrowable](
+    offsets: Map[Partition, Offset],
+    commitF: NonEmptyMap[TopicPartition, OffsetAndMetadata] => G[Unit],
+  ): G[Boolean] = {
+    offsets.toNem.fold(Applicative[G].pure(false)) { offsetsNem =>
+      commitF(
+        offsetsNem
+          .mapBoth {
+            case (partition, offset) => TopicPartition(topic, partition) -> OffsetAndMetadata(offset, commitMetadata)
+          },
+      ).as(true)
+    }
+  }
+}
+
+private[replicator] object AsyncPeriodicTopicCommit {
+
+  def make[F[_]: Temporal](
+    topic: Topic,
+    commitMetadata: Metadata,
+    commitPeriod: FiniteDuration,
+    consumer: KafkaConsumer[F, ?, ?],
+    log: Log[F],
+  ): Resource[F, TopicCommit[F]] = {
+    val mkInstanceF: F[AsyncPeriodicTopicCommit[F]] = for {
+      stateRef <- Ref.of[F, State](State())
+    } yield new AsyncPeriodicTopicCommit[F](
+      topic          = topic,
+      commitMetadata = commitMetadata,
+      consumer       = consumer,
+      stateRef       = stateRef,
+      log            = log,
+    )
+
+    for {
+      instance <- mkInstanceF.toResource
+
+      periodicCommitProcessF = (
+        Temporal[F].sleep(commitPeriod) >>
+          instance
+            .periodicCommitStep()
+            .recoverWith {
+              case t: Throwable =>
+                log.error(s"periodic offset commit failed: ${t.getMessage}", t)
+            }
+      ).foreverM[Unit]
+
+      _ <- periodicCommitProcessF.background
+    } yield instance
+  }
+
+  // opened to the package for tests
+  private[commit] final case class State(
+    assignedPartitions: Map[Partition, PartitionState] = Map.empty,
+  ) {
+    def addAssignedPartitions(partitions: Iterable[Partition]): State = {
+      copy(
+        assignedPartitions = partitions.foldLeft(assignedPartitions) { (prevAssigned, partition) =>
+          prevAssigned.updatedWith(partition)(prevStateOpt => prevStateOpt.getOrElse(PartitionState()).some)
+        },
+      )
+    }
+
+    def removeAssignedPartitions(partitions: Iterable[Partition]): State = {
```

Comment:
maybe:
```suggestion
    def removeRevokedPartitions(partitions: Iterable[Partition]): State = {
```

Reply:
The state doesn't know anything about "revoked" or "lost" partitions - it just has a set of "assigned" partitions. I don't want to pollute it semantically with complications from higher layers here - i.e. lost vs revoked etc.

Reply:
"removing assigned partitions" feels wrong too - then maybe better to have two functions:
* `addPartitions` or `trackPartitions`
* `removePartitions`?

Reply:
it could be `unassign` but current naming looks OK to me.
- `assignedPartitions` in the state
- `addAssignedPartitions` and `removeAssignedPartitions` for manipulating the set